lib/agent0/agent0/hyperdrive/interactive/interactive_hyperdrive_agent.py

```diff
@@ -0,0 +1,208 @@
+
+
+class InteractiveHyperdriveAgent:
+    """Interactive Hyperdrive Agent.
+    This class is barebones with documentation, will just call the corresponding function
+    in the interactive hyperdrive class to keep all logic in the same place. Adding these
+    wrappers here for ease of use.
+    """
+
+    def __init__(self, base: FixedPoint, eth: FixedPoint, name: str | None, pool: InteractiveHyperdrive):
+        """Constructor for the interactive hyperdrive agent.
+        NOTE: this constructor shouldn't be called directly, but rather from InteractiveHyperdrive's
+        `init_agent` method.
+
+        Arguments
+        ---------
+        base: FixedPoint
+            The amount of base to fund the agent with.
+        eth: FixedPoint
+            The amount of ETH to fund the agent with.
+        name: str
+            The name of the agent. Defaults to the wallet address.
+        pool: InteractiveHyperdrive
+            The pool object that this agent belongs to.
+        """
+        self._pool = pool
+        self.name = name
+        # TODO
+        self.agent = self._pool._init_agent(base, eth, name)
+
+    @property
+    def wallet(self) -> HyperdriveWallet:
+        """Returns the agent's current wallet.
+
+        Returns
+        -------
+        HyperdriveWallet
+            The agent's current wallet.
+        """
+        return self.agent.wallet
+
+    def add_funds(self, base: FixedPoint | None = None, eth: FixedPoint | None = None) -> None:
+        """Adds additional funds to the agent.
+
+        Arguments
+        ---------
+        base : FixedPoint
+            The amount of base to fund the agent with. Defaults to 0.
+        eth : FixedPoint
+            The amount of ETH to fund the agent with. Defaults to 0.
+        """
+        if base is None:
+            base = FixedPoint(0)
+        if eth is None:
+            eth = FixedPoint(0)
+        self._pool._add_funds(self.agent, base, eth)
+
+    def create_checkpoint(self, checkpoint_time: int | None = None) -> CreateCheckpoint:
+        """Creates a checkpoint, originating from the agent.
+
+        Arguments
+        ---------
+        checkpoint_time : int
+            The checkpoint time to mint
+
+        Returns
+        -------
+        CreateCheckpoint
+            The emitted event of the checkpoint call.
+        """
+        # TODO underlying function will raise NotImplementedError, fix
+        return self._pool._create_checkpoint(self.agent, checkpoint_time)
+
+    def open_long(self, base: FixedPoint) -> OpenLong:
+        """Opens a long for this agent.
+
+        Arguments
+        ---------
+        base : FixedPoint
+            The amount of longs to open in units of base.
+
+        Returns
+        -------
+        OpenLong
+            The emitted event of the open long call.
+        """
+        return self._pool._open_long(self.agent, base)
+
+    def close_long(self, maturity_time: int, bonds: FixedPoint) -> CloseLong:
```

Comment:
This should be in units of base. But, also:

After an wonderful journey through `InteractiveHyperdrive -> async_execute_agent_trades -> async_execute_single_agent_trade -> async_match_contract_call_to_trade -> HyperdriveInterface` you will find that we end up actually calling the `api::_contract_calls::_async_close_long` function via `HyperdriveInterface.async_close_long`. We are passing that interface in as an argument way back at the top of this stack, when we know that it's an individual `close_long` trade. I think the chosen path is so that we can get access to all of the error handling. But that leads me to wonder if we should be taking that path directly from the API in the first place?

If you think it is easy to fix this wiring go ahead, otherwise please make an issue for it.

Reply:
[#1107](https://github.com/delvtech/agent0/issues/1107)